lib/scholar/covariance/ledoit_wolf.ex

```diff
@@ -0,0 +1,352 @@
+  end
+
+  defnp ledoit_wolf(x, opts) do
+    case Nx.shape(x) do
+      {_n, 1} ->
+        {
+          Nx.pow(x, 2)
+          |> Nx.mean()
+          |> Nx.broadcast({1, 1}),
+          0.0
+        }
+
+      _ ->
+        ledoit_wolf_complex(x, opts)
+    end
+  end
+
+  defnp ledoit_wolf_complex(x, opts) do
+    n_features = Nx.axis_size(x, 1)
+    shrinkage = ledoit_wolf_shrinkage(x, opts)
+    emp_cov = empirical_covariance(x, opts)
+    # git
+    mu = Nx.sum(trace(emp_cov)) / n_features
+    shrunk_cov = (1.0 - shrinkage) * emp_cov
+    mask = Nx.iota(Nx.shape(shrunk_cov))
+    selector = Nx.remainder(mask, n_features + 1) == 0
+    shrunk_cov = Nx.select(selector, shrunk_cov + shrinkage * mu, shrunk_cov)
+    {shrunk_cov, shrinkage}
+  end
+
+  defnp empirical_covariance(x, _opts) do
+    n = Nx.axis_size(x, 0)
+
+    covariance =
+      Nx.transpose(x)
+      |> Nx.dot(x)
+      |> Nx.divide(n)
+
+    case Nx.shape(covariance) do
+      {} -> Nx.reshape(covariance, {1, 1})
+      _ -> covariance
+    end
+  end
+
+  defnp trace(x) do
+    n = Nx.axis_size(x, 0)
+
+    Nx.eye(n)
+    |> Nx.multiply(x)
+    |> Nx.sum()
+  end
+
+  defnp ledoit_wolf_shrinkage(x, opts) do
+    case Nx.shape(x) do
+      {_, 1} ->
+        0
+
+      {n} ->
+        Nx.broadcast(x, {1, n})
+        |> ledoit_wolf_shrinkage_complex(opts)
+
+      _ ->
+        ledoit_wolf_shrinkage_complex(x, opts)
+    end
+  end
+
+  defnp ledoit_wolf_shrinkage_complex(x, opts) do
+    {n_samples, n_features} = Nx.shape(x)
+    {_, size} = Nx.type(x)
+    block_size = opts[:block_size]
+    n_splits = (n_features / block_size) |> Nx.floor() |> Nx.as_type({:s, size})
+
+    x2 = Nx.pow(x, 2)
+    emp_cov_trace = Nx.sum(x2, axes: [0]) / n_samples
+    mu = Nx.sum(emp_cov_trace) / n_features
+    beta = Nx.tensor(0.0, type: {:f, size})
+    delta = Nx.tensor(0.0, type: {:f, size})
+    i = Nx.tensor(0)
+    block = Nx.iota({block_size})
+
+    if n_splits > 0 do
+      {beta, delta, _} =
+        while {beta, delta, {x, x2, block, i, n_splits}}, i < n_splits do
+          {block_size} = Nx.shape(block)
+          j = Nx.tensor(0)
+
+          {beta, delta, _} =
+            while {beta, delta, {x, x2, block, i, j, n_splits}}, j < n_splits do
+              {block_size} = Nx.shape(block)
+              rows_from = block_size * i
+              cols_from = block_size * j
+              x2_t = Nx.transpose(x2)
+              x_t = Nx.transpose(x)
+
+              to_beta =
+                Nx.slice_along_axis(x2_t, rows_from, block_size, axis: 0)
+                |> Nx.dot(Nx.slice_along_axis(x2, cols_from, block_size, axis: 1))
+                |> Nx.sum()
+
+              beta = beta + to_beta
+
+              to_delta =
+                Nx.slice_along_axis(x_t, rows_from, block_size, axis: 0)
+                |> Nx.dot(Nx.slice_along_axis(x, cols_from, block_size, axis: 1))
+                |> Nx.pow(2)
+                |> Nx.sum()
```

Comment:
```suggestion
                Nx.slice_along_axis(x, rows_from, block_size, axis: 1)
                |> Nx.dot([0], Nx.slice_along_axis(x, cols_from, block_size, axis: 1), [0])
                |> Nx.pow(2)
                |> Nx.sum()
```

Reply:
this Nx.pow(2) here seems odd, mathematically speaking. is it correct?

Reply:
I’ll double-check that, but at first glance - it looks correct.

Reply:
So, if we take the [scikit-learn implementation](https://github.com/scikit-learn/scikit-learn/blob/6e9039160f0dfc3153643143af4cfdca941d2045/sklearn/covariance/_shrunk_covariance.py#L376) as a reference, then it is correct. But I have to agree with you - this looks weird. I don't see a reason to implement it that way - I will simplify it in the next commit.

Reply:
I will simplify this whole section significantly in the next commit. After longer investigation of the scikit-learn code, I found it highly overcomplicated for Scholar. Maybe in Python it has some reason to calculate beta and delta in the loop, but I don't believe this applies here.